grpcbp/server.go

```diff
@@ -0,0 +1,148 @@
+	// connection would be closed by sending a GoAway. Idleness duration is
+	// defined since the most recent time the number of outstanding RPCs became
+	// zero or the connection establishment. The current default value is
+	// infinity.
+	MaxConnectionIdle time.Duration `yaml:"maxConnectionIdle"`
+
+	// MaxConnectionAge is a duration for the maximum amount of time a
+	// connection may exist before it will be closed by sending a GoAway. A
+	// random jitter of +/-10% will be added to MaxConnectionAge to spread out
+	// connection storms. The current default value is infinity.
+	MaxConnectionAge time.Duration `yaml:"maxConnectionAge"`
+
+	// MaxConnectionAgeGrace is an additive period after MaxConnectionAge after
+	// which the connection will be forcibly closed. The current defualt value is
+	// infinity.
+	MaxConnectionAgeGrace time.Duration `yaml:"maxConnectionAgeGrace"`
+
+	// After a duration of this time if the server doesn't see any activity it
+	// pings the client to see if the transport is still alive. If set below 1s,
+	// a minimum value of 1s will be used instead. The current default value is 2
+	// hours.
+	PingInactiveClientsAfter time.Duration `yaml:"pingInactiveClientsAfter"`
+
+	// After having pinged for keepalive check, the server waits for a duration
+	// of KeepAliveTimeout and if no activity is seen even after that the connection is
+	// closed. The current default value is 20 seconds.
+	KeepAliveTimeout time.Duration `yaml:"keepAliveTimeout"`
+
+	// KeepAliveMinTime is the minimum amount of time a client should wait before sending
+	// a keepalive ping.
+	KeepAliveMinTime time.Duration `yaml:"keepAliveMinTime"`
+
+	// If true, server allows keepalive pings even when there are no active
+	// streams(RPCs). If false, and client sends ping when there are no active
+	// streams, server will send GOAWAY and close the connection.
+	KeepAlivePermitWithoutStream bool `yaml:"keepAlivePermitWithoutStream"`
+}
+
+// Validate ensures all set values are within bound.
+func (s ServerConfig) Validate() error {
+	return nil
+}
+
+// GRPCServer provides passing in the generated gRPC service implementation and
+// reigster it on the created gRPC.Server.
+type GRPCServer interface {
+	RegisterOn(*grpc.Server)
+}
+
+// NewBaseplateServer returns a new gRPC implementation of a Baseplate server
+// with the given config.
+func NewBaseplateServer(bp baseplate.Baseplate, server GRPCServer, cfg ServerConfig) (baseplate.Server, error) {
+	lis, err := net.Listen("tcp", bp.GetConfig().Addr)
+	if err != nil {
+		return nil, err
+	}
+	grpcServer, err := NewServer(bp, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	server.RegisterOn(grpcServer)
+
+	return ApplyBaseplate(bp, grpcServer, lis), nil
+}
+
+// NewServer returns a new instance of grpc.Server with any baseplate-related
+// server options.
+func NewServer(bp baseplate.Baseplate, cfg ServerConfig) (*grpc.Server, error) {
```

Comment:
maybe it's my lack of context around grpc, but `NewBaseplateServer` vs. `NewServer` confuses me here, even with the function doc.

is it that we are supposed to first call `grpcbp.NewServer` then pass that into `grpcbp.NewBaseplateServer`?

Reply:
I'll have to defer to what @kylelemons had in mind. The way I changed it made it more confusing. The goal here is to offer a way for the service-specific code to register their methods. For example for this here https://grpc.io/docs/languages/go/basics/#starting-the-server it would be:

```
pb.RegisterRouteGuideServer(grpcServer, newServer())
```
and since baseplate.go takes care of instantiating `*grpc.Server` we need to expose this instantiated type somehow so that the baseplate user can register their concrete types. Previously I had this in the server config, another way could look like this:

```go
func NewBaseplateServer(bp baseplate.Baseplate, registerFunc func(*grpc.Server), cfg ServerConfig) (baseplate.Server, error) {
	lis, err := net.Listen("tcp", bp.GetConfig().Addr)
	if err != nil {
		return nil, err
	}
	srv, err := NewServer(bp, cfg)
	if err != nil {
		return nil, err
	}

        registerFunc(srv)

	return ApplyBaseplate(bp, srv, lis), nil
}
```
But I believe Kyle was thinking about some specific using an interface type :slightly_smiling_face: